public/sw.js

```diff
@@ -0,0 +1,231 @@
+const OFFLINE_VERSION = 1;
+const OFFLINE_CACHE_NAME = "offline-cache-v1";
+const DATA_CACHE_NAME = "data-cache-v1";
+const NOTIFICATION_CACHE_NAME = "notification-cache-v1";
+const OFFLINE_URL = "offline.html";
+const assets = [
+  "/",
+  "/offline.html",
+  "/manifest.json",
+  "/static/fonts/montserrat.css",
+  "/logo192.png",
+  "/favicon.ico",
+];
+let sessionId = null;
+let notificationTimer = null;
+
+self.addEventListener("message", async (event) => {
+  if (event.data.sessionId) {
+    sessionId = event.data.sessionId;
+    const cache = await caches.open(NOTIFICATION_CACHE_NAME);
+    const response = new Response(JSON.stringify(event.data.timestamp));
+    await cache.put("lastInteractionTime", response);
+    console.log("Session ID updated in service worker:", sessionId);
+  }
+  startPollingNotifications();
+});
+
+function startPollingNotifications() {
+  if (notificationTimer !== null) {
+    console.log("Polling already running.");
+    return;
+  }
+  pollNotifications();
+}
+
+async function pollNotifications() {
+  if (!sessionId || Notification.permission !== "granted") {
+    return;
+  }
+
+  const canSendNotification = await shouldSendNotification();
+  if (!canSendNotification) {
+    console.log("Notification was recently shown. Skipping notification.");
+    notificationTimer = setTimeout(pollNotifications, 5000);
+    return;
+  }
+  try {
+    // Temporary json file/endpoint until I've received new endpoint
+    // const response = await fetch(`https://api.zeeguu.org/available_languages`);
+    const response = await fetch(`/mock_notification.json`);
+    if (response.ok) {
+      const data = await response.json();
+      const randomNumber = Math.floor(Math.random() * 4);
+      const notificationUrl = data[randomNumber].url
+        ? data[randomNumber].url
```

Comment:
This was just made for the 4 mock notifications I used for testing via. the mock_notification.json file. However, as this won't be used, it shouldn't matter as it will be removed. The url was made, so that I could exemplify that we could make the notifications redirect the user to a specific URL. However, if no url was given, it would simply sent the user to the front page.